contributor-book/src/getting-started/testing.md

```diff
@@ -0,0 +1,19 @@
+# testing
+
+## Test for TensorOps
+
+the following examples use matrix multiplication operation
+
+Test for Tensor operations (as in given this input, expect it match or approximate this output) are defined only in [`burn-tensor/src/test/ops`](https://github.com/tracel-ai/burn/blob/b9bd42959b0d3e755a25e383cb5b38beb25559b8/burn-tensor/src/tests/ops/matmul.rs#L1) and not in the backends, with the exception of `burn-autodiff`. These test are added to the `testgen_all` macro rule in [`burn-tensor/src/test/mod.rs`](https://github.com/tracel-ai/burn/blob/b9bd42959b0d3e755a25e383cb5b38beb25559b8/burn-tensor/src/tests/mod.rs#L59). This is then propagated to the existing backends without any additional work.
```

Comment:
That's hard to read.

Also I would recommend to avoid pointing to lines of code because the code base will evolve and we will end up with broken links. I suggest to point to files only and then use struct names to point a specific part in the file.

Reply:
actually, all the links are permalinks, as in line numbers on specific commits. Those should only be changed if the code undergoes changes so significant that the specific section would need to be rewritten

Reply:
Also what do you mean it's hard to read? do you mean the structure of the text is hard to understand or the unrendered markdown links are hard to read?

Reply:
I had to reread it several times but I was not using the rendered files indeed 😆. While those links are locked to a specific commit I believe it is better to make the documentation more resilient to the changes by avoiding the necessity to link to the code base.

Also a trick to avoid inlining HTTP links in the paragraphs is to write them at the end of the file. They become `[the burn-tensor crate]()` in the text and then at the end of the file we have `[the burn-tensor crate]: https://...`. This link can be also reused easily in the same file making it easier to update.

Here is an example of how we can avoid being to close the source code:

> Tensor operations tests are defined in the `burn-tensor` crate under the `test` folder.  This crate defines a macro called `testgen_all` which generates the code for the common tensor operations tests in all the backends crates like `burn-wgpu`, `burn-ndarray`, etc...

Reply:
> I had to reread it several times but I was not using the rendered files indeed 😆. While those links are locked to a specific commit I believe it is better to make the documentation more resilient to the changes by avoiding the necessity to link to the code base.

for sections like the architecture, that's probably a good idea, though in the case of  guides for contributions that involve a significant amount of changes across the project  I disagree. take the case of the op addition. Some of the files linked are a thousand+ lines, and may involve changes in multiple places. Plus, IMHO it's generally better to lean towards making instructions as explicit and unambiguous as possible. It's easy to scan over a file and miss something, I definitely did a few times.

> Also a trick to avoid inlining HTTP links in the paragraphs is to write them at the end of the file. They become [the burn-tensor crate]() in the text and then at the end of the file we have [the burn-tensor crate]: https://.... This link can also be reused easily in the same file making it easier to update.

I'm currently using that for the footnotes, generally for useful information or tips not directly related to burn.

Reply:
Fine with me. But we should avoid linking to the code base when it is not strictly necessary. For instance I don't think it is really necessary to link to the exact line of the macro rule definition in the file. In a big file there can be value to those links indeed.